pkg/uploader/provider/kopia.go

```diff
@@ -0,0 +1,204 @@
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	repokeys "github.com/vmware-tanzu/velero/pkg/repository/keys"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo"
+	"github.com/vmware-tanzu/velero/pkg/repository/udmrepo/service"
+)
+
+//BackupFunc mainly used to make testing more convenient
+var BackupFunc = kopia.Backup
+var RestoreFunc = kopia.Restore
+
+//kopiaProvider recorded info related with kopiaProvider
+type kopiaProvider struct {
+	bkRepo        udmrepo.BackupRepo
+	credGetter    *credentials.CredentialGetter
+	uploader      *snapshotfs.Uploader
+	restoreCancel chan struct{}
+	log           logrus.FieldLogger
+}
+
+//NewKopiaUploaderProvider initialized with open or create a repository
+func NewKopiaUploaderProvider(
+	ctx context.Context,
+	credGetter *credentials.CredentialGetter,
+	bsl *velerov1api.BackupStorageLocation,
+	log logrus.FieldLogger,
+) (Provider, error) {
+	kp := &kopiaProvider{
+		log:        log,
+		credGetter: credGetter,
+	}
+	//repoUID which is used to generate kopia repository config with unique directory path
+	repoUID := string(bsl.GetUID())
+	repoOpt, err := udmrepo.NewRepoOptions(
+		udmrepo.WithPassword(kp, ""),
+		udmrepo.WithConfigFile("", repoUID),
+		udmrepo.WithDescription("Initial kopia uploader provider"),
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error to get repo options")
+	}
+
+	repoSvc := service.Create(log)
+	log.WithField("repoUID", repoUID).Info("Opening backup repo")
+
+	kp.bkRepo, err = repoSvc.Open(ctx, *repoOpt)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to find kopia repository")
+	}
+	return kp, nil
+}
+
+//CheckContext check context status periodically
+//check if context is timeout or cancel
+func (kp *kopiaProvider) CheckContext(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			if kp.uploader != nil {
+				kp.uploader.Cancel()
+				kp.log.Infof("Backup is been canceled")
+			}
+			if kp.restoreCancel != nil {
+				close(kp.restoreCancel)
+				kp.log.Infof("Restore is been canceled")
+			}
+			return
+		default:
+			time.Sleep(time.Second * 10)
+		}
+	}
+}
+
+func (kp *kopiaProvider) Close(ctx context.Context) {
+	kp.bkRepo.Close(ctx)
+}
+
+//RunBackup which will backup specific path and update backup progress
+func (kp *kopiaProvider) RunBackup(
+	ctx context.Context,
+	path string,
+	tags map[string]string,
+	parentSnapshot string,
+	updater uploader.ProgressUpdater) (string, error) {
+	if updater == nil {
+		return "", errors.New("Need to initial backup progress updater first")
+	}
+
+	log := kp.log.WithFields(logrus.Fields{
+		"path":           path,
+		"parentSnapshot": parentSnapshot,
+	})
+	repoWriter := kopia.NewShimRepo(kp.bkRepo)
+	kp.uploader = snapshotfs.NewUploader(repoWriter)
+	prorgess := new(kopia.KopiaProgress)
+	prorgess.InitThrottle(backupProgressCheckInterval)
+	prorgess.Updater = updater
+	kp.uploader.Progress = prorgess
```

Comment:
Seems the `Progress` of the shared `uploader` will be overwritten when calling Runbackups() 
 more than once.

Reply:
yes, I've removed the uploader from the kopiaProvider struct. which will make the progress is independent for every backup